external-crates/move/crates/move-compiler/src/expansion/translate.rs

```diff
@@ -2717,65 +2755,100 @@ fn exp_(context: &mut Context, sp!(loc, pe_): P::Exp) -> E::Exp {
+                    let pkg = context.current_package;
+                    context.env().check_feature(FeatureGate::DotCall, pkg, loc);
+                    let tys_opt = optional_types(context, ptys_opt);
+                    let ers = sp(rloc, exps(context, prs));
+                    EE::MethodCall(edotted, n, is_macro, tys_opt, ers)
+                }
+                None => {
+                    assert!(context.env().has_errors());
+                    EE::UnresolvedError
+                }
             }
-        },
-        PE::Cast(e, ty) => EE::Cast(exp(context, *e), type_(context, ty)),
+        }
+        PE::Cast(e, ty) => EE::Cast(exp(context, e), type_(context, ty)),
         PE::Index(..) => {
             // TODO index syntax will be added
             context.spec_deprecated(loc, /* is_error */ true);
             EE::UnresolvedError
         }
-        PE::Annotate(e, ty) => EE::Annotate(exp(context, *e), type_(context, ty)),
+        PE::Annotate(e, ty) => EE::Annotate(exp(context, e), type_(context, ty)),
         PE::Spec(_) => {
             context.spec_deprecated(loc, /* is_error */ false);
             EE::Unit { trailing: false }
         }
         PE::UnresolvedError => EE::UnresolvedError,
     };
-    sp(loc, e_)
+    Box::new(sp(loc, e_))
 }
 
-// If the expression is a named block, hand back the name and a normal block. Otherwise, just
-// process the expression. This is used to lift names for loop and while to the appropriate form.
-fn maybe_named_loop(
+// If the expression can take a label, attach the label. Otherwise error
+fn maybe_labeled_exp(
     context: &mut Context,
     loc: Loc,
-    body: P::Exp,
-) -> (Option<BlockLabel>, Box<E::Exp>) {
-    if let P::Exp_::NamedBlock(name, seq) = body.value {
-        (
-            Some(name),
-            Box::new(sp(loc, E::Exp_::Block(sequence(context, loc, seq)))),
-        )
-    } else {
-        (None, exp(context, body))
+    label: BlockLabel,
+    e: Box<E::Exp>,
+) -> Box<E::Exp> {
+    let sp!(_eloc, e_) = *e;
+    let e_ = match e_ {
+        E::Exp_::While(label_opt, cond, body) => {
+            unique_label(context, loc, &label, label_opt);
+            E::Exp_::While(Some(label), cond, body)
+        }
+        E::Exp_::Loop(label_opt, body) => {
+            unique_label(context, loc, &label, label_opt);
+            E::Exp_::Loop(Some(label), body)
+        }
+        E::Exp_::Block(label_opt, seq) => {
+            unique_label(context, loc, &label, label_opt);
+            E::Exp_::Block(Some(label), seq)
+        }
+        _ => {
+            let msg = format!(
+                "Invalid label. Labels can only be used on 'while', 'loop', or block '{{}}' \
+                 expressions"
+            );
+            context
+                .env()
+                .add_diag(diag!(Syntax::InvalidLabel, (loc, msg)));
+            E::Exp_::UnresolvedError
+        }
+    };
+    Box::new(sp(loc, e_))
+}
+
+fn unique_label(
```

Comment:
Nit: rename `ensure_unique_label` or similar just to help understand the code above.